### PR TITLE
merkle: return hashes for empty merkle trees

### LINKED ISSUE #5192: Merkle tree is not compatible with rfc6962 for empty sets
@liamsi discovered that our Merkle tree implementation which is supposed to be compatible with RFC6962 diverges for empty tress. This is true for both Go and Rust, which are also currently incompatible with eachother - https://github.com/informalsystems/tendermint-rs/issues/498

According to RFC6962, empty trees should use the hash of the empty string as their root, while in Tendermint we just use the empty string.

This means eg. blocks with no transactions have an empty value for the DataHash in the header. One possible benefit of this is that it means empty blocks have smaller headers, which is nice, but it's not clear that benefit is worth diverging from the spec.

The only hashes in the header that this would really effect are the DataHash and the EvidenceHash, which are both often empty. So this change would make Headers 64 bytes bigger for empty blocks, 32 bytes bigger for blocks with transactions, and the same size as blocks with transactions and evidence. Of course we might be able to remove the evidence hash down the road circa #2530 

We might want to consider fixing this for v0.34 ...

### PR DESCRIPTION
**Pull request by erikgrinaker**
Fixes #5192.

@liamsi Can you verify that the test vectors match the Rust implementation? I updated `ProofsFromByteSlices()` as well, anything else that should be updated?

### CODE REVIEW

File: crypto/merkle/tree.go [hunk 1]
```diff
@@ -2,14 +2,16 @@ package merkle
...
 )
 
 // HashFromByteSlices computes a Merkle tree where the leaves are the byte slice,
-// in the provided order.
+// in the provided order. It follows RFC-6962.
 func HashFromByteSlices(items [][]byte) []byte {
 	switch len(items) {
 	case 0:
-		return nil
+		return tmhash.Sum([]byte{})
```
Comment (liamsi) [inline crypto/merkle/tree.go:14]: It would probably be more consistent with the existing code if there was a function `func emptyHash() []byte` where the other cases are defined too (`leafHash` and `innerHash`)

File: crypto/merkle/tree_test.go [hunk 2]
```diff
@@ -17,16 +19,45 @@ func (tI testItem) Hash() []byte {
...
+			assert.Equal(t, tc.expectHash, hex.EncodeToString(hash))
+		})
+	}
+}
+
 func TestProof(t *testing.T) {
 
+	// Try an empty proof first
+	rootHash, proofs := ProofsFromByteSlices([][]byte{})
+	require.Equal(t, "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", hex.EncodeToString(rootHash))
```
Comment (liamsi) [inline crypto/merkle/tree_test.go:50]: ✅

File: crypto/merkle/rfc6962_test.go [hunk 3]
```diff
@@ -28,13 +28,20 @@ func TestRFC6962Hasher(t *testing.T) {
...
 		got  []byte
 		want string
 	}{
-		// Since creating a merkle tree of no leaves is unsupported here, we skip
-		// the corresponding trillian test vector.
+		// Check that empty trees return the hash of an empty string.
+		// echo -n '' | sha256sum
+		{
+			desc: "RFC6962 Empty Tree",
+			want: "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"[:tmhash.Size*2],
```
Comment (liamsi) [inline crypto/merkle/rfc6962_test.go:42]: why times two in `[:tmhash.Size*2]`.
Matches sha256({}): https://github.com/google/trillian/blob/7166f485ebc965d24116db667c9e058a8ed9275c/merkle/rfc6962/rfc6962_test.go#L39

**Comment by liamsi** [comment]
> Can you verify that the test vectors match the Rust implementation?

Not yet because the rust implementation treats the case even differently (it returns [all zeroes](https://github.com/informalsystems/tendermint-rs/blob/99e8b9db738d7aec61126b2c9519f6818e8ee875/tendermint/src/merkle.rs#L24)) 🙈 and we don't have a test-case for the empty-tree yet (in fact test-cases are only just a few): 
https://github.com/informalsystems/tendermint-rs/blob/99e8b9db738d7aec61126b2c9519f6818e8ee875/tendermint/src/merkle.rs#L97-L129

I'll submit a PR to the Rust implementation that and fixes and adds this case (and generally a few more test-cases).

>  I updated ProofsFromByteSlices() as well, anything else that should be updated?

I don't think anything else besides failing tests need updating. It would be advisable to get a really good understanding if there are places where we ever could have hashed an all empty slice (do the failing tests represent sth that could have happened in practice too?). 

While this change looks really innocent, it's effectively like changing the underlying hash function which is a breaking change.

**Comment by erikgrinaker** [comment]
> It would be advisable to get a really good understanding if there are places where we ever could have hashed an all empty slice

The main worry would be whether this affects hardware devices, do you know?

File: crypto/merkle/tree.go (hunk 1 above)
Comment (erikgrinaker) [reply crypto/merkle/tree.go]: Agreed, done

**Comment by liamsi** [comment]
Hmm, good question. I don't think it does affect hardware devices but I'm not 100% certain. I checked the cosmos ledger app and there it just takes a message and hashes it before signing (so the software on the ledger won't need any changes): https://github.com/cosmos/ledger-cosmos/blob/a511b01b419669873567aef724bcfa1cf802d0e2/app/src/crypto.c#L74-L81
And for the KMS: as long as this does not change the "sign bytes" (the test-vectors that @marbar3778 recently reintroduced and extended (?)), it should not have any affect either.

**Comment by jleni** [comment]
I am not sure about the context but.. In both apps, before signing any message, the app will parse it and validate.
In the cosmos app, it needs to be a valid JSON with required fields, etc. In the tendermint app, it will expect amino (protobuf?) and needs to find round, height, etc.  
An empty message will get rejected, it will never get hashed or signed.

**Comment by erikgrinaker** [comment]
>An empty message will get rejected, it will never get hashed or signed.

This affects anything that is hashed, not just transactions. Consider e.g. `EvidenceHash` in the block header. Most blocks do not contain evidence, so this field is currently empty, but will contain `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 ` after this change.

Hashes may also be used as inputs to other hashes (the SDK e.g. does this with their multi-store), such that the final hash will be different from what it is today.

**Comment by jleni** [comment]
Yes, my answer was in the context of hardware wallets (Ledger Nano S and X).
I don't expect any of those additional cases to affect hardware wallet implementations in any way.

**Comment by jleni** [comment]
The only other hashes (though I doubt you will affect those) that we use... are during address generation.. but I expect those to remain unaffected.. otherwise, this would be a much more deeper issue

**Comment by erikgrinaker** [comment]
I've updated the tests - only a single test vector failure, which was expected (`Message_BlockResponse`), so I think this should be fine. Thanks for the update on hardware @jleni!

There are some cases that are a bit unclear. For example, `Block.Hash()` and `Header.Hash()` return `nil` if there is no block or header; this is probably fine. Otherwise we'll need to do a thorough pass to update instances like these and their effects.

Would appreciate reviews, and thoughts on potential adverse effects.

**Comment by erikgrinaker** [comment]
> should we add a short exercpt to the upgrading.md

Ah, forgot about that - done, thanks!

**Comment by erikgrinaker** [comment]
@ebuchman @liamsi After a quick review of the code, I have found the following hashes which return `nil` rather than the RFC6962 empty hash when they are empty:

* `Block.Hash()`
* `PartSet.Hash()`
* `Header.Hash()`
* `Commit.Hash()`
* `Evidence.Hash()`
* `ValidatorSet.Hash()`

My sense is that we should update these to return RFC6962 empty hashes as well, so I'll go ahead and port them, but would appreciate some input here. This will e.g. cause header fields like `last_commit.block_id.hash` and `last_commit.block_id.parts.hash` to contain a hash for the initial block.

**Comment by liamsi** [comment]
> @liamsi Can you verify that the test vectors match the Rust implementation? I updated ProofsFromByteSlices() as well, anything else that should be updated?

All existing test-cases in the rust code base were not affected by updating this on the rust side too (due to the lack of tests). I'll add a test-case there for empty trees too.

https://github.com/informalsystems/tendermint-rs/pull/514
